src/Recurrence.ts

```diff
@@ -212,4 +207,149 @@ export class Recurrence {
 
         return this.toText() === other.toText(); // this also checks baseOnToday
     }
+
+    /**
+     * nextAfter returns the next occurrence's date after `after`, based on the given rrule.
+     *
+     * The common case is that `rrule.after` calculates the next date and it
+     * can be used as is.
+     *
+     * In the special cases of monthly and yearly recurrences, there exists an
+     * edge case where an occurrence after the given number of months or years
+     * is not possible. For example: A task is due on 2022-01-31 and has a
+     * recurrence of `every month`. When marking the task as done, the next
+     * occurrence will happen on 2022-03-31. The reason being that February
+     * does not have 31 days, yet RRule sets `bymonthday` to `31` for lack of
+     * having a better alternative.
+     *
+     * In order to fix this, `after` will move into the past day by day. Each
+     * day, the next occurrence is checked to be after the given number of
+     * months or years. By moving `after` into the past day by day, it will
+     * eventually calculate the next occurrence based on `2022-01-28`, ending up
+     * in February as the user would expect.
+     */
+    private nextAfter(after: Moment, rrule: RRule): Date {
+        // We need to remove the timezone, as rrule does not regard timezones and always
+        // calculates in UTC.
+        // The timezone is added again before returning the next date.
+        after.utc(true);
+        let next = window.moment(rrule.after(after.toDate()));
+
+        // If this is a monthly recurrence, treat it special.
+        const monthMatch = this.toText().match(/every( \d+)? month(s)?(.*)?/);
+        if (monthMatch !== null) {
+            // ... unless the rule fixes the date, such as 'every month on the 31st'
+            if (!this.toText().includes(' on ')) {
+                next = Recurrence.nextAfterMonths(after, next, rrule, monthMatch[1]);
+            }
+        }
+
+        // If this is a yearly recurrence, treat it special.
+        const yearMatch = this.toText().match(/every( \d+)? year(s)?(.*)?/);
+        if (yearMatch !== null) {
+            next = Recurrence.nextAfterYears(after, next, rrule, yearMatch[1]);
```

Comment:
Don't we need the same check for ` on ` here? :thinking:

Reply:
Yes, I wondered about that. Will add a failing test and then do it.

Reply:
I tried to write a failing test for this, and couldn't get it to fail.

Here's the test I wrote: eac9896680f6d89eaf5d314fced613559c6283cb

It just works, so I have left the implementation code unchanged.